corehq/apps/receiverwrapper/rate_limiter.py

```diff
@@ -102,19 +106,39 @@ def report_submission_usage(domain):
     _report_current_global_submission_thresholds()
 
 
-def _delay_and_report_rate_limit_submission(domain, max_wait, datadog_metric):
+def _delay_and_report_rate_limit_submission(domain, max_wait, delay_rather_than_reject, datadog_metric):
+    """
+    Attempt to acquire permission from the rate limiter waiting up to 15 seconds.
+
+    When delay_rather_than_reject is False
+
+        If it's acquired, report throttle_method:delay and duration:<bucketed duration>;
+        otherwise report throttle_method:reject and duration:delayed_reject or quick_reject,
+        depending on whether the rate limiter bothered to wait or could tell there was no chance.
+
+    When delay_rather_than_reject is True
+
+        If it's acquired, report throttle_method:delay and duration:<bucketed duration> (as before);
+        otherwise report throttle_method:delay and duration:delay_rather_than_reject
+
+    Returns whether the permission was eventually acquired (with no variation on delay_rather_than_reject).
+    """
     with TimingContext() as timer:
         acquired = submission_rate_limiter.wait(domain, timeout=max_wait)
     if acquired:
         duration_tag = bucket_value(timer.duration, [.5, 1, 5, 10, 15], unit='s')
+    elif delay_rather_than_reject:
+        if timer.duration < max_wait:
+            time.sleep(max_wait - timer.duration)
```

Comment:
Is this a good idea? Seems like it could cause a balloon in gevent workers, which may have negative memory consequences.

Reply:
That's a reasonable question. The `submission_rate_limiter.wait` command a few lines above will use time.sleep in sub-second increments in a loop to keep checking up for up to `max_wait` which defaults to 15 seconds, and that's been live for quite some time (since Dec 2019, I believe). It's certainly possible that that causes problems that haven't bubbled up into something we've notice, but I haven't seen it.

This change makes `submit_form_locally` sleep for up to 100 ms (`.1` below) if a "real" submission from that domain would be throttled/rejected. I picked 100 ms because it's on the same order of magnitude as the minimum amount of time we could expect a form submission to take, and so would less-than-double the total amount of time needed to get through N such submissions even if it was being throttled the whole time. The biggest users of `submit_form_locally`, I believe are in case imports and auto case updates, which are both in celery processes that don't use `gevent`, though certainly other places might.

Anyway those are my thoughts on that. Curious to hear what you think.

Reply:
Sounds like you've thought it through. My only remaining concern is we are unlikely to know if this poses a real problem until it gets really bad. Will we remember to look here if this causes the problem I described?

Is `time.sleep` patched to `gevent.sleep` in the context of Gunicorn? If not, should we use `gevent.sleep` instead?

Reply:
My understanding is that `gevent.monkey.patch_all()` patches `time.sleep` by default, and we use that default. I double checked the code where we do that and the definition of the `patch_all` function, and it seems to bear that out as well.

I think the potential issue you're describing is mainly a concern on webworkers, if there was some high-volume route that called this function synchronously (as opposed to through celery). We have a metric on our HQ Vitals datadog dashboard that tracks the percent of worker processes in use (split out by mobile_webworkers vs hq_webworkers) so we'd definitely see if it was reaching it's max. We calibrate the machines/process counts so that we basically never hit (or even get that close to) the limit, so it would certainly stand out if it did.

I also in one of my later commits put the "new" part of this PR, rate limiting via a delay within `submit_form_locally` by default, under a flag. That way if there is an issue with that, it's something we can turn off quickly. (There was a pre-existing similar delay in case imports, which I rewrote to go through `submit_form_locally` instead. That's a change in implementation rather than a change in functionality, and I didn't put that change under the flag.)